File: lib/data/dto/wallet_dto.dart
class WalletDto {
  final int totalPoints;
  final int todayPoints;
  final int todayReferralPoints;
  final int uncollectedPoints;
  final int allTimePointsEarned;

  const WalletDto({
    required this.totalPoints,
    required this.todayPoints,
    required this.todayReferralPoints,
    required this.uncollectedPoints,
    required this.allTimePointsEarned,
  });

  factory WalletDto.fromJson(Map<String, dynamic> json) => WalletDto(
        totalPoints: (json['totalPoints'] as num? ??
                json['balance'] as num? ??
                0)
            .toInt(),
        todayPoints: (json['todayPoints'] as num?)?.toInt() ?? 0,
        todayReferralPoints:
            (json['todayReferralPoints'] as num?)?.toInt() ?? 0,
        uncollectedPoints:
            (json['uncollectedPoints'] as num?)?.toInt() ?? 0,
        allTimePointsEarned:
            (json['allTimePointsEarned'] as num?)?.toInt() ?? 0,
      );
}

class TransactionDto {
  final String id;
  final String description;
  final int points;
  final DateTime timestamp;
  final TransactionTypeDto type;

  const TransactionDto({
    required this.id,
    required this.description,
    required this.points,
    required this.timestamp,
    required this.type,
  });

  factory TransactionDto.fromJson(Map<String, dynamic> json) => TransactionDto(
        id: (json['id'] ?? json['_id'] ?? '').toString(),
        description: (json['description'] ?? json['reason'] ?? '').toString(),
        points: (json['points'] as num? ?? json['amount'] as num? ?? 0).toInt(),
        timestamp: json['timestamp'] != null
            ? DateTime.tryParse(json['timestamp'].toString()) ?? DateTime.now()
            : (json['createdAt'] != null
                ? DateTime.tryParse(json['createdAt'].toString()) ??
                    DateTime.now()
                : DateTime.now()),
        type: TransactionTypeDto.fromString(
            (json['type'] ?? 'screen_time').toString()),
      );
}

enum TransactionTypeDto {
  screenTime('screen_time'),
  pointsCollected('points_collected'),
  spinWheel('spin_wheel'),
  game('game'),
  task('task'),
  survey('survey'),
  referral('referral'),
  giftCardPurchase('gift_card_purchase'),
  cashOut('cash_out'),
  donation('donation'),
  raffleWin('raffle_win'),
  welcomeBonus('welcome_bonus');

  final String value;
  const TransactionTypeDto(this.value);

  static TransactionTypeDto fromString(String s) =>
      TransactionTypeDto.values.firstWhere(
        (e) => e.value == s,
        orElse: () => TransactionTypeDto.screenTime,
      );
}

File: backend/src/wallet/dto/wallet.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { IsOptional, IsEnum } from 'class-validator';
import { PaginationQueryDto, PaginationMeta } from '../../common/dto/pagination.dto';

export enum TransactionType {
  SCREEN_TIME = 'screen_time',
  POINTS_COLLECTED = 'points_collected',
  SPIN_WHEEL = 'spin_wheel',
  GAME = 'game',
  TASK = 'task',
  SURVEY = 'survey',
  REFERRAL = 'referral',
  GIFT_CARD_PURCHASE = 'gift_card_purchase',
  CASH_OUT = 'cash_out',
  DONATION = 'donation',
  RAFFLE_WIN = 'raffle_win',
}

export class WalletDto {
  @ApiProperty({ description: 'Total points in wallet (available to spend)' })
  totalPoints: number;

  @ApiProperty({ description: 'Points collected today' })
  todayPoints: number;

  @ApiProperty({ description: 'Referral points earned today' })
  todayReferralPoints: number;

  @ApiProperty({ description: 'Uncollected points waiting to be claimed (capped)' })
  uncollectedPoints: number;

  @ApiProperty({ description: 'All-time total points earned (sum of all positive transactions)' })
  allTimePointsEarned: number;
}

export class TransactionDto {
  @ApiProperty()
  id: string;

  @ApiProperty()
  description: string;

  @ApiProperty({ description: 'Positive = earned, negative = spent' })
  points: number;

  @ApiProperty()
  timestamp: Date;

  @ApiProperty({ enum: TransactionType })
  type: TransactionType;
}

export class TransactionQueryDto extends PaginationQueryDto {
  @ApiPropertyOptional({ enum: TransactionType, description: 'Filter by transaction type' })
  @IsOptional()
  @IsEnum(TransactionType)
  type?: TransactionType;
}

export class PaginatedTransactionsDto {
  @ApiProperty({ type: [TransactionDto] })
  data: TransactionDto[];

  @ApiProperty()
  meta: PaginationMeta;
}
